File: src/NetFxCSharp/Controls/ADNullableValueControls/ADNullableValueControls.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace ActiveDev.Controls
{
    [CLSCompliant(true)]
    public class ADNullableValueControls : Collection<IADNullableValueControl>
    {
        public static ADNullableValueControls FromContainerControl(Control cControl)
        {
            return FromContainerControlInternal(new ADNullableValueControls(), cControl);
        }

        private static ADNullableValueControls FromContainerControlInternal(ADNullableValueControls nullableControls, Control cControl)
        {
            foreach (Control c in cControl.Controls)
            {
                if (c.GetType().GetInterface(typeof(IADNullableValueControl).Name) != null)
                {
                    nullableControls.Add(((IADNullableValueControl)c));
                }

                if (c.HasChildren)
                {
                    FromContainerControlInternal(nullableControls, ((Control)c));
                }
            }

            return nullableControls;
        }

        public string CheckForNotAllowedNullValues()
        {
            string locString = "";
            foreach (IADNullableValueControl ic in this)
            {
                if (ic.NullValueMessage != null)
                {
                    if (ic.NullValueMessage != "")
                    {
                        if (ic.Value.IsNull)
                        {
                            locString += "* " + ic.Text + " " + ic.NullValueMessage + System.Environment.NewLine + System.Environment.NewLine;
                        }
                    }
                }
            }

            if (locString == "")
            {
                return null;
            }
            else
            {
                return locString;
            }
        }
    }
}

File: src/NetFxCSharp/Controls/ADNullableValueControls/ADNullableValueControls.vb
Imports System.Windows.Forms
Imports System.Collections.ObjectModel

<CLSCompliant(True)> _
Public Class ADNullableValueControls
    Inherits Collection(Of IADNullableValueControl)

    Public Shared Function FromContainerControl(ByVal cControl As Control) As ADNullableValueControls
        Return FromContainerControlInternal(New ADNullableValueControls, cControl)
    End Function

    Private Shared Function FromContainerControlInternal(ByVal nullableControls As ADNullableValueControls, ByVal cControl As Control) As ADNullableValueControls

        For Each c As Control In cControl.Controls
            If c.GetType.GetInterface(GetType(IADNullableValueControl).Name) IsNot Nothing Then
                nullableControls.Add(DirectCast(c, IADNullableValueControl))
            End If
            If c.HasChildren Then
                FromContainerControlInternal(nullableControls, DirectCast(c, Control))
            End If
        Next
        Return nullableControls
    End Function

    Public Function CheckForNotAllowedNullValues() As String

        Dim locString As String = ""

        For Each ic As IADNullableValueControl In Me
            If ic.NullValueMessage IsNot Nothing Then
                If ic.NullValueMessage <> "" Then
                    If ic.Value.IsNull Then
                        locString &= "* " & ic.Text & " " & ic.NullValueMessage & vbNewLine & vbNewLine
                    End If
                End If
            End If
        Next
        If locString = "" Then
            Return Nothing
        Else
            Return locString
        End If
    End Function
End Class


